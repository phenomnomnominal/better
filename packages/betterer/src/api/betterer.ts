import type { BettererSuiteSummary } from '../suite/index.js';
import type { BettererOptions } from './types.js';

import { BettererRunnerΩ } from '../runner/index.js';
import { merge } from './merge.js';
import { results } from './results.js';
import { runner } from './runner.js';
import { watch } from './watch.js';

/**
 * @public Run **Betterer** with the given options.
 *
 * @example
 * ```typescript
 * import { betterer } from '@betterer/betterer';
 *
 * const suiteSummary = await betterer(options);
 * ```
 *
 * @param options - Options for running **Betterer**.
 * @throws {@link @betterer/errors#BettererError | `BettererError` }
 * Will throw if something goes wrong while running **Betterer**.
 */
export const betterer = async function betterer(options: BettererOptions = {}): Promise<BettererSuiteSummary> {
  const runner = await BettererRunnerΩ.create(options);
  return await runner.run();
};

betterer.merge = merge;
betterer.results = results;
betterer.runner = runner;
betterer.watch = watch;
