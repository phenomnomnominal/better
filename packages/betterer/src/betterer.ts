import { debug } from '@phenomnomnominal/debug';

import { BettererBaseConfigPartial, BettererStartConfigPartial, BettererWatchConfigPartial } from './config';
import { BettererSummary } from './context';
import { BettererRunner, BettererRunnerΩ, BettererWatcherΩ } from './runner';

export async function betterer(partialConfig: BettererStartConfigPartial = {}): Promise<BettererSummary> {
  initDebug();
  const runner = new BettererRunnerΩ();
  await runner.start(partialConfig);
  await runner.queue();
  return runner.stop();
}

export async function runner(partialConfig?: BettererBaseConfigPartial): Promise<BettererRunner> {
  initDebug();
  const runner = new BettererRunnerΩ();
  await runner.start(partialConfig);
  return runner;
}
betterer.runner = runner;

export async function watch(partialConfig: BettererWatchConfigPartial = {}): Promise<BettererRunner> {
  initDebug();
  partialConfig.watch = true;
  const runner = new BettererRunnerΩ();
  const context = await runner.start(partialConfig);
  const watcher = new BettererWatcherΩ(context, runner);
  await watcher.setup();
  return watcher;
}
betterer.watch = watch;

function initDebug(): void {
  const enabled = !!process.env.BETTERER_DEBUG;
  if (enabled) {
    debug({
      header: 'betterer',
      include: [/@betterer\//],
      ignore: [require.resolve('./utils')],
      enabled,
      time: !!process.env.BETTERER_DEBUG_TIME,
      values: !!process.env.BETTERER_DEBUG_VALUES,
      logPath: process.env.BETTERER_DEBUG_LOG
    });
  }
}
