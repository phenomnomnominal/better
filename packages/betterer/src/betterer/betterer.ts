import logDiff from 'jest-diff';

import { BettererGoal, BettererGoalFunction, BettererTest, BettererConstraint, BettererDiff } from './types';

export type BettererOptions<TestType, SerialisedType = TestType> = {
  test: BettererTest<TestType>;
  constraint: BettererConstraint<SerialisedType>;
  goal: BettererGoal<SerialisedType>;
  diff: BettererDiff<TestType, SerialisedType>;
  isOnly?: boolean;
  isSkipped?: boolean;
};

export class Betterer<Base = unknown, Serialised = Base> {
  public constraint: BettererConstraint<Serialised>;
  public diff: BettererDiff<Base, Serialised>;
  public goal: BettererGoalFunction<Serialised>;
  public test: BettererTest<Base>;
  public isOnly = false;
  public isSkipped = false;
  public name = '';

  constructor({
    test,
    constraint,
    goal,
    diff = defaultDiff,
    isOnly = false,
    isSkipped = false
  }: BettererOptions<Base, Serialised>) {
    this.constraint = constraint;
    this.diff = diff;
    this.goal = createGoal(goal);
    this.test = test;
    this.isOnly = isOnly;
    this.isSkipped = isSkipped;
  }

  public only(): Betterer<Base, Serialised> {
    this.isOnly = true;
    return this;
  }

  public skip(): Betterer<Base, Serialised> {
    this.isSkipped = true;
    return this;
  }
}

function createGoal(goal: BettererGoal<unknown>): BettererGoalFunction<unknown> {
  if (typeof goal === 'function') {
    return goal as BettererGoalFunction<unknown>;
  }
  return (value: unknown): boolean => value === goal;
}

function defaultDiff(_: unknown, serialisedCurrent: unknown, serialisedPrevious: unknown): void {
  const diffStr =
    logDiff(serialisedPrevious, serialisedCurrent, {
      aAnnotation: 'Previous',
      bAnnotation: 'Current'
    }) || '';
  console.log(
    diffStr
      .split('\n')
      .map((line) => `  ${line}`)
      .join('\n')
  );
}
