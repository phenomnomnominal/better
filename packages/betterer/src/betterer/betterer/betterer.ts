import { BettererRun } from '../../context';
import { BettererConstraint, BettererGoal, BettererTestFunction, BettererOptions } from './types';

export class Betterer<TestType = unknown> {
  public readonly constraint: BettererConstraint<TestType>;
  public readonly goal: BettererGoal<TestType>;
  public readonly test: BettererTestFunction<TestType>;

  public readonly isBetterer = true;

  private _isOnly: boolean;
  private _isSkipped: boolean;

  constructor(options: BettererOptions<TestType>) {
    const { constraint, test, isOnly, isSkipped } = options;
    this.constraint = constraint;
    this.goal = this._createGoal(options);
    this.test = test;
    this._isOnly = isOnly || false;
    this._isSkipped = isSkipped || false;
  }

  public get isOnly(): boolean {
    return this._isOnly;
  }

  public get isSkipped(): boolean {
    return this._isSkipped;
  }

  public only(): this {
    this._isOnly = true;
    return this;
  }

  public skip(): this {
    this._isSkipped = true;
    return this;
  }

  public getExpected(run: BettererRun): TestType {
    return run.test.expected as TestType;
  }

  private _createGoal(options: BettererOptions<TestType>): BettererGoal<TestType> {
    const hasGoal = Object.hasOwnProperty.call(options, 'goal');
    if (!hasGoal) {
      return (): boolean => false;
    }
    const { goal } = options;
    if (this._isGoalFunction(goal)) {
      return goal;
    }
    return (value: unknown): boolean => value === goal;
  }

  private _isGoalFunction(goal: unknown): goal is BettererGoal<unknown> {
    return typeof goal === 'function';
  }
}

export function isBetterer(obj: unknown): obj is Betterer {
  return !!(obj as Betterer).isBetterer;
}
