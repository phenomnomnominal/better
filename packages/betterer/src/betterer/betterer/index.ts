export * from './betterer';
export * from './types';
