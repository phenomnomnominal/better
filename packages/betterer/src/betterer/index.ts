export { createBetterer } from './create-betterer';
export { Betterer, isBetterer } from './betterer';
export { BettererFiles, BettererFile, FileBetterer, createFileBetterer, isFileBetterer } from './file-betterer';
export { SerialisableBetterer, isSerialisableBetterer } from './serialisable-betterer';
export * from './betterer/types';
export * from './file-betterer/types';
export * from './serialisable-betterer/types';
export * from './types';
