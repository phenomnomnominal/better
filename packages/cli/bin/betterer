#!/usr/bin/env node
import { cli__ } from '../dist/cli.js';
cli__(process.cwd(), process.argv);
