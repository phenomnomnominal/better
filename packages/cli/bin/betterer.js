#!/usr/bin/env node
import { cliΔ } from '../dist/cli.js';
void cliΔ(globalThis.process.cwd(), globalThis.process.argv);
