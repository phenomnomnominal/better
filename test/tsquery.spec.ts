import { betterer } from '@betterer/betterer';

import { createFixture } from './fixture';

describe('betterer', () => {
  it('should report the existence of TSQuery matches', async () => {
    const { logs, paths, readFile, cleanup, resolve, writeFile, runNames } = await createFixture('tsquery', {
      '.betterer.ts': `
import { tsquery } from '@betterer/tsquery';

export default {
  tsquery: () => tsquery(
    'CallExpression > PropertyAccessExpression[expression.name="console"][name.name="log"]'
  ).include('./src/**/*.ts')
};
    `
    });

    const configPaths = [paths.config];
    const resultsPath = paths.results;
    const indexPath = resolve('./src/index.ts');

    await writeFile(indexPath, `console.log('foo');`);

    const newTestRun = await betterer({ configPaths, resultsPath, workers: 1 });

    expect(runNames(newTestRun.new)).toEqual(['tsquery']);

    const sameTestRun = await betterer({ configPaths, resultsPath, workers: 1 });

    expect(runNames(sameTestRun.same)).toEqual(['tsquery']);

    await writeFile(indexPath, `console.log('foo');\nconsole.log('foo');`);

    const worseTestRun = await betterer({ configPaths, resultsPath, workers: 1 });

    expect(runNames(worseTestRun.worse)).toEqual(['tsquery']);

    const result = await readFile(resultsPath);

    expect(result).toMatchSnapshot();

    await writeFile(indexPath, ``);

    const betterTestRun = await betterer({ configPaths, resultsPath, workers: 1 });

    expect(runNames(betterTestRun.better)).toEqual(['tsquery']);

    const completedTestRun = await betterer({ configPaths, resultsPath, workers: 1 });

    expect(runNames(completedTestRun.completed)).toEqual(['tsquery']);

    expect(logs).toMatchSnapshot();

    await cleanup();
  });
});
