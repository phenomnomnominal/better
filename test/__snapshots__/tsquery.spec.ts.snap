// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`betterer should report the existence of TSQuery matches 1`] = `
"// BETTERER RESULTS V2.
//
// If this file contains merge conflicts, use \`betterer merge\` to automatically resolve them:
// https://phenomnomnominal.github.io/betterer/docs/results-file/#merge
//
exports[\`tsquery\`] = {
  value: \`{
    \\"src/index.ts:4087252068\\": [
      [0, 0, 11, \\"TSQuery match\\", \\"3870399096\\"]
    ]
  }\`
};
"
`;

exports[`betterer should report the existence of TSQuery matches 2`] = `
Array [
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 tsquery: running \\"tsquery\\"!
",
  "🌟 Betterer (0ms): 1 test running...
✅ tsquery: \\"tsquery\\" got checked for the first time! (1 issue) 🎉
",
  "🎉 Betterer (0ms): 1 test done!
✅ tsquery: \\"tsquery\\" got checked for the first time! (1 issue) 🎉
",
  "🎉 Betterer (0ms): 1 test done!
✅ tsquery: \\"tsquery\\" got checked for the first time! (1 issue) 🎉

1 test got checked. 🤔
1 test got checked for the first time! 🎉
",
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 tsquery: running \\"tsquery\\"!
",
  "🌟 Betterer (0ms): 1 test running...
✅ tsquery: \\"tsquery\\" stayed the same. (1 issue) 😐
",
  "🎉 Betterer (0ms): 1 test done!
✅ tsquery: \\"tsquery\\" stayed the same. (1 issue) 😐
",
  "🎉 Betterer (0ms): 1 test done!
✅ tsquery: \\"tsquery\\" stayed the same. (1 issue) 😐

1 test got checked. 🤔
1 test stayed the same. 😐
",
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 tsquery: running \\"tsquery\\"!
",
  "🌟 Betterer (0ms): 1 test running...
🔥 tsquery: \\"tsquery\\" got worse. (1 new issue, 2 total) 😔
",
  "🌟 Betterer (0ms): 1 test running...
🔥 tsquery: \\"tsquery\\" got worse. (1 new issue, 2 total) 😔

Error: \\"tsquery\\" got worse. (1 new issue, 2 total) 😔
",
  "💥 Betterer (0ms): 1 test done! 1 test errored!
🔥 tsquery: \\"tsquery\\" got worse. (1 new issue, 2 total) 😔
・ 1 existing issue in \\"<project>/fixtures/tsquery/src/index.ts\\".
・ New issue in \\"<project>/fixtures/tsquery/src/index.ts\\"!
・
・   <project>/fixtures/tsquery/src/index.ts
・   1 | console.log('foo');
・ > 2 | console.log('foo');
・     | ^^^^^^^^^^^ TSQuery match
・

Error: \\"tsquery\\" got worse. (1 new issue, 2 total) 😔
",
  "💥 Betterer (0ms): 1 test done! 1 test errored!
🔥 tsquery: \\"tsquery\\" got worse. (1 new issue, 2 total) 😔
・ 1 existing issue in \\"<project>/fixtures/tsquery/src/index.ts\\".
・ New issue in \\"<project>/fixtures/tsquery/src/index.ts\\"!
・
・   <project>/fixtures/tsquery/src/index.ts
・   1 | console.log('foo');
・ > 2 | console.log('foo');
・     | ^^^^^^^^^^^ TSQuery match
・

Error: \\"tsquery\\" got worse. (1 new issue, 2 total) 😔

1 test got checked. 🤔
1 test got worse. 😔

You should try to fix the new issues! As a last resort, you can run \`betterer --update\` to force an update of the results file. 🆙
",
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 tsquery: running \\"tsquery\\"!
",
  "🌟 Betterer (0ms): 1 test running...
✅ tsquery: \\"tsquery\\" met its goal! 🎉
",
  "🎉 Betterer (0ms): 1 test done!
✅ tsquery: \\"tsquery\\" met its goal! 🎉
",
  "🎉 Betterer (0ms): 1 test done!
✅ tsquery: \\"tsquery\\" met its goal! 🎉

1 test got checked. 🤔
1 test got better! 😍
1 test met its goal! 🎉
",
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 tsquery: running \\"tsquery\\"!
",
  "🌟 Betterer (0ms): 1 test running...
✅ tsquery: \\"tsquery\\" has already met its goal! ✨
",
  "🎉 Betterer (0ms): 1 test done!
✅ tsquery: \\"tsquery\\" has already met its goal! ✨
",
  "🎉 Betterer (0ms): 1 test done!
✅ tsquery: \\"tsquery\\" has already met its goal! ✨

1 test got checked. 🤔
1 test met its goal! 🎉
",
]
`;
